Add log to metrics plugin

Athishpranav2003 opened the pull request:
<!--
Thank you for contributing to Fluent Operator!
Your commits need to follow DCO: https://probot.github.io/apps/dco/
And please provide the following information to help us make the most of your pull request:
-->

### What this PR does / why we need it:
Add `LogToMetrics` Filter plugin
### Which issue(s) this PR fixes:
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #1285 

### Does this PR introduced a user-facing change?
Nope
<!--
If no, just write "None" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md
-->
```release-note
Add `LogToMetrics` Filter plugin to Fluentbit component 
```

### Additional documentation, usage docs, etc.:
Updated in the codebase itself.
<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->

Athishpranav2003 commented:
@cosmo0920 the build is failing for some other reason. I have raised PR so that we can get started with the discussions. Meanwhile i will try to look at the build issue. Also running make is actually modifying some other files dont know why. Would need your help in this part.

Do we need to add tests for this plugin?

benjaminhuo commented on apis/fluentbit/v1alpha2/plugins/filter/log_to_metrics_types.go:
```diff
@@ -0,0 +1,112 @@
+	// Value Format: FIELD REGEX
+	Exclude []string `json:"exclude,omitempty"`
+	// Defines the mode for the metric. Valid values are [counter, gauge or histogram]
+	Metric_mode string `json:"metric_mode,omitempty"`
```
```suggestion
	MetricMode string `json:"metricMode,omitempty"`
```

benjaminhuo replied:
shouldn't use _

benjaminhuo commented on apis/fluentbit/v1alpha2/plugins/filter/log_to_metrics_types.go:
```diff
@@ -0,0 +1,112 @@
+	// Sets a help text for the metric.
+	Metric_description string `json:"metric_description,omitempty"`
+	// Defines a bucket for histogram
+	Bucket []float `json:"bucket,omitempty"`
```
pls check the github action errors like below:
![image](https://github.com/user-attachments/assets/f886d8c1-1ad2-4cc6-ae79-5ac06c26aad1)

benjaminhuo replied:
https://github.com/fluent/fluent-operator/actions/runs/10398584230/job/28798064743?pr=1305

Athishpranav2003 replied:
Sure, I tried testing locally but there were different errors. I will check and update the PR

cosmo0920 replied:
Should use `[]string` type here. This is because the config map type is same as `regex` parameter, isn't it?

Athishpranav2003 replied:
Oh ok

Athishpranav2003 commented:
@benjaminhuo Now the local build is working. I guess make is autoformatting the code and wantedly changing 
```bash
        modified:   apis/fluentbit/v1alpha2/plugins/filter/zz_generated.deepcopy.go
	modified:   apis/fluentbit/v1alpha2/zz_generated.deepcopy.go
	modified:   go.mod
	modified:   go.sum
```
Not sure why. I removed those changes and then pushed it for now

Athishpranav2003 commented:
@benjaminhuo push to ghcr seems to fail (403 error code). I guess thats issue with some configuration in gcp i guess. Does this seem fine now?

Athishpranav2003 commented:
@cosmo0920 Could you check the PR? i  have addressed your comments

benjaminhuo commented:
> Not sure why. I removed those changes and then pushed it for now

The auto-generated files shouldn't be removed.
You should run `make manifests generate fmt vet docs-update` or even `make test docs-update` and commit everything generated @Athishpranav2003

Athishpranav2003 commented:
Ok fine
Will do it and update

Athishpranav2003 commented:
@benjaminhuo Have added them too as well

cosmo0920 commented:
I put an approve and kicked CI tasks as well.

Athishpranav2003 commented:
@cosmo0920 seems like its passing the tests. Anything else i missed?

benjaminhuo commented on apis/fluentbit/v1alpha2/plugins/filter/log_to_metrics_types.go:
```diff
@@ -0,0 +1,112 @@
+	// Sets the name of the metric.
+	MetricName string `json:"metric_name,omitempty"`
+	// Namespace of the metric
+	MetricNamespace string `json:"metric_namespace,omitempty"`
```
all the json field should be `camelCase`:
metric_namespace -> metricNamespace
metric_subsystem -> metricSubsystem

etc.

Athishpranav2003 commented:
@benjaminhuo addressed your comments. I wasnt sure about the convention part that you guys followed

benjaminhuo commented:
> @benjaminhuo addressed your comments. I wasnt sure about the convention part that you guys followed

Thanks for the contribution! @Athishpranav2003
